cpp/ql/src/semmle/code/cpp/ir/internal/TranslatedExpr.qll

```diff
@@ -1866,12 +1917,295 @@ abstract class TranslatedCall extends TranslatedNonConstantExpr {
+      tag = AllocationExtentConvertTag() and opcode instanceof Opcode::Convert or
+      tag = AllocationElementSizeTag() and opcode instanceof Opcode::Constant or
+      tag = AllocationSizeTag() and opcode instanceof Opcode::Mul // REVIEW: Overflow?
+    )
+  }
+
+  override final Instruction getInstructionSuccessor(InstructionTag tag,
+      EdgeKind kind) {
+    kind instanceof GotoEdge and
+    (
+      (
+        tag = AllocationExtentConvertTag() and
+        result = getInstruction(AllocationElementSizeTag())
+      ) or
+      (
+        tag = AllocationElementSizeTag() and
+        result = getInstruction(AllocationSizeTag())
+      ) or
+      (
+        tag = AllocationSizeTag() and
+        result = getParent().getChildSuccessor(this)
+      )
+    )
+  }
+
+  override final TranslatedElement getChild(int id) {
+    id = 0 and result = getExtent()
+  }
+
+  override final Instruction getChildSuccessor(TranslatedElement child) {
+    child = getExtent() and
+    result = getInstruction(AllocationExtentConvertTag())
+  }
+
+  override final string getInstructionConstantValue(InstructionTag tag) {
+    tag = AllocationElementSizeTag() and
+    result = newArrayExpr.getAllocatedElementType().getSize().toString()
+  }
+
+  override final Instruction getInstructionOperand(InstructionTag tag,
+      OperandTag operandTag) {
+    (
+      tag = AllocationSizeTag() and
+      (
+        operandTag instanceof LeftOperand and result = getInstruction(AllocationExtentConvertTag()) or
+        operandTag instanceof RightOperand and result = getInstruction(AllocationElementSizeTag())
+      )
+    ) or
+    (
+      tag = AllocationExtentConvertTag() and
+      operandTag instanceof UnaryOperand and
+      result = getExtent().getResult()
+    )
+  }
+
+  private TranslatedExpr getExtent() {
+    result = getTranslatedExpr(newArrayExpr.getExtent().getFullyConverted())
+  }
+}
+
+/**
+ * The IR translation of a call to `operator new` as part of a `new` or `new[]`
+ * expression.
+ */
+class TranslatedAllocatorCall extends TTranslatedAllocatorCall,
+    TranslatedCall {
+  NewOrNewArrayExpr newExpr;
+
+  TranslatedAllocatorCall() {
+    this = TTranslatedAllocatorCall(newExpr) and
+    expr = newExpr
+  }
+
+  override final string toString() {
+    result = "Allocator call for " + newExpr.toString()
+  }
+
+  override final predicate producesExprResult() {
+    none()
+  }
+
+  override final Instruction getFirstCallTargetInstruction() {
+    result = getInstruction(CallTargetTag())
+  }
+
+  override final Instruction getCallTargetResult() {
+    result = getInstruction(CallTargetTag())
+  }
+
+  override predicate hasInstruction(Opcode opcode, InstructionTag tag,
+      Type resultType, boolean isGLValue) {
+    TranslatedCall.super.hasInstruction(opcode, tag, resultType, isGLValue) or
+    (
+      tag = CallTargetTag() and
+      opcode instanceof Opcode::FunctionAddress and
+      resultType instanceof BoolType and //HACK
```

Comment:
Please elaborate on what the problem is and why `BoolType` is the best choice here. Is it because `void*` is not always in a db? We do have `void` in every db, so maybe use that and set `isGLValue = true`.

Reply:
Changed to `glval<Unknown>`. I didn't want to use `void` because there's no such thing as a glvalue of type `void` in C++, and checking for `instr.getResultType() instanceof VoidType` is the current way to determine if an instruction returns a result at all.